### DISCUSSION TITLE
Keeping west modules in sync with main tree

### DISCUSSION THREAD

**Discussion by nashif** [zephyrproject-rtos/zephyr#28651] [General] [answered]
How do I keep my west modules updates across rebases, bisecting and other git operations when doing upstream development and when using the upstream manifest.

**Comment by nashif** [comment, marked as answer]
add two git hooks:

```
$ cat  .git/hooks/post-merge
#!/bin/sh

CHANGED=`git diff HEAD@{1} --stat -- west.yml | wc -l`
if [ $CHANGED -gt 0 ];
then
    echo "west.yml has changed!"
    west update
fi
```

and
 
```
$ cat  .git/hooks/post-checkout
#!/bin/sh

GIT_ROOT=`git rev-parse --show-toplevel`
CHANGED=`git diff HEAD@{1} --stat -- ${GIT_ROOT}/west.yml | wc -l`
if [ $CHANGED -gt 0 ];
then
    echo "west.yml has changed!"
    west update
fi
```


Now every time you move in git history, modules will get updated and be in sync with main git repo.

**Comment by mbolivar-nordic** [comment]
Be careful with this advice. It only works if zephyr is your manifest repository.

It does not work in other situations, such as if you're importing the zephyr manifest file somewhere else, or if your manifest file uses projects with revisions set to branches.

**Reply by nashif** [reply]
> Be careful with this advice. It only works if zephyr is your manifest repository.

yeah, we are in the zephyr project where the manifest is the default manifest. There is no single answer for every possible configuration out there and you would expect people to look at the script and see this for them selves. I have been using this method for almost a year now, it works like a charm.

**Reply by mbolivar-nordic** [reply]
>  you would expect people to look at the script and see this for them selves.

I would definitely not expect that. People copy/paste things blindly all the time.

> I have been using this method for almost a year now, it works like a charm.

Good for you! I think my cautionary comment still stands. It seems like you're just using upstream Zephyr with zephyr/west.yml as your manifest, and I'm glad that suits your use cases. I expect that many of the users I care about are doing something else.

**Reply by nashif** [reply]
> People copy/paste things blindly all the time.

Sure, but if that does not work, people will come back here and complain or they will eventually figure it out on their own.

>  I expect that many of the users I care about are doing something else.

Am I missing something? Where in the zephyr project do we have manifests other than in the main repo? If you have downstream projects and repos that do things differently, you can provide guidance on how to do that there with the correct paths and working scripts. Saying that this does not work for your users so the majority of zephyr users should not use it is absurd.

**Reply by mbolivar-nordic** [reply]
>  Saying that this does not work for your users so the majority of zephyr users should not use it is absurd.

Please be so kind as to point out where I said that?

**Reply by nashif** [reply]
> Please be so kind as to point out where I said that?

`I expect that many of the users I care about are doing something else.`

**Reply by mbolivar-nordic** [reply]
There is an implicit assumption about what the majority of zephyr users are doing in your argument.

**Reply by nashif** [reply]
the question is 
> How do I keep my west modules updates across rebases, bisecting and other git operations when doing upstream development and when using the upstream manifest.

so, there are no assumptions at all, this is not for everyone, this is for someone doing upstream development dealing with rebases and bisecting on a regular basis..

I am not sure to be honest what the argument here is about, this is a common issue and this is one solution, it works, if you have something better to suggest, please go ahead and be my guest.

**Reply by marc-hb** [reply]
This looks like a violent agreement. Good advice but think twice and don't blindly copy/paste. Thank you both :-)
